{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LincKer\n",
    "This programme identifies neighbouring mRNAs for a list of genes of interest (in this case, lncRNAs) in a GFF file. For each lncRNA, the programme locates up to 5 upstream and downstream neighbours, although this parameter can be adjusted according to user requirements. The output includes the distance between the lncRNA and each neighbouring mRNA, the differentially expressed (DE) status of neighbouring mRNAs, and other relevant information. The files needed to run the notebooks can be found in the notebooks_files folder.\n",
    "\n",
    "*Executed with Python v3.11.4*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![Open in Google Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/yahelGB/M.Sc_thesis/blob/main/notebooks/lncker.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import os\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialise a dictionary to store information for each contig\n",
    "contigs = {}\n",
    "\n",
    "# load and read GFF file and extract features\n",
    "with open('/path/to/GCF_003789085.1_ASM378908v1_genomic.gff', 'r') as f_input:\n",
    "    for line in f_input:\n",
    "        if line.startswith('#') or not line.strip():\n",
    "            continue\n",
    "\n",
    "        # parse the line into parts using the tab delimiter\n",
    "        parts = line.strip().split('\\t')\n",
    "        contig_id = parts[0]\n",
    "        feature_type = parts[2]\n",
    "        start = int(parts[3])\n",
    "        end = int(parts[4])\n",
    "        strand = parts[6]\n",
    "        metadata = parts[8]\n",
    "        \n",
    "        # structuring metadata in a dictionary\n",
    "        metadata_split = metadata.split(';')\n",
    "        metadata_dict = {item.split('=')[0]: item.split('=')[1] for item in metadata_split if '=' in item}\n",
    "\n",
    "        # create a feature dictionary\n",
    "        feature = {\n",
    "            'type': feature_type,\n",
    "            'start': start,\n",
    "            'end': end,\n",
    "            'strand': strand,\n",
    "            'metadata': metadata_dict,\n",
    "        }\n",
    "\n",
    "        # Add the feature to the corresponding contig\n",
    "        if contig_id not in contigs:\n",
    "            contigs[contig_id] = []\n",
    "        contigs[contig_id].append(feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify the path to the file containing the IDs of interest\n",
    "# note: update 'input_filename' with the correct path and filename each time \n",
    "# a new set of lncRNAs is analysed, these files are generated by the notebook \n",
    "# 'exclusive_lncRNAs.ipynb'.\n",
    "\n",
    "input_filename = '/path/to/sample_exclusive_lncRNAs.txt'\n",
    "ids_interes = set()\n",
    "\n",
    "# read the IDs from the specified file and store them in a set\n",
    "with open(input_filename, 'r') as f_ids:\n",
    "    for line in f_ids:\n",
    "        ids_interes.add(line.strip())\n",
    "\n",
    "print('Number of lncRNAs loaded:', len(ids_interes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the differentially expressed mRNAs and store them in a set\n",
    "de_mrnas = set()\n",
    "with open('/path/to/de_mRNAs.txt', 'r') as f_de:\n",
    "    for line in f_de:\n",
    "        de_mrnas.add(line.strip())\n",
    "\n",
    "print('Number of mRNAs DE:', len(de_mrnas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary to store the contigs containing the IDs of interest\n",
    "contigs_interes = defaultdict(list)\n",
    "\n",
    "# check each contig for IDs of interest\n",
    "for contig_id, features in contigs.items():\n",
    "    for feature in features:\n",
    "        if feature['type'] == 'lnc_RNA':\n",
    "            metadata_dict = feature['metadata']\n",
    "            gene_id = metadata_dict.get('ID', 'Unknown')\n",
    "            if gene_id.startswith('rna-'):\n",
    "                gene_id = gene_id[4:]\n",
    "            if gene_id in ids_interes:\n",
    "                contigs_interes[gene_id].append(contig_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "# main function \n",
    "def find_multiple_mRNA_neighbors(features, lncRNA_index, num_neighbors=5):\n",
    "    \"\"\"\n",
    "    search for mRNA neighbours around a given lncRNA, by default, identifies\n",
    "    up to 5 upstream and 5 downstream neighbours, though the number of neighbours\n",
    "    can be adjusted according to user requirements.\n",
    "\n",
    "    only the first mRNA after each gene is considered, ignoring mRNA variants.\n",
    "    \"\"\"\n",
    "    upstream_mRNAs = []\n",
    "    downstream_mRNAs = []\n",
    "    lncRNA = features[lncRNA_index]\n",
    "\n",
    "    # neighbouring gene tracking\n",
    "    last_gene_upstream = None\n",
    "    last_gene_downstream = None\n",
    "\n",
    "    # search up to 5 upstream neighbours\n",
    "    for i in range(lncRNA_index - 1, -1, -1):\n",
    "        feature = features[i]\n",
    "        if feature['type'] == 'gene':\n",
    "            last_gene_upstream = i\n",
    "        elif feature['type'] == 'mRNA' and last_gene_upstream is not None:\n",
    "            upstream_mRNAs.append(i)\n",
    "            last_gene_upstream = None\n",
    "            if len(upstream_mRNAs) == num_neighbors:\n",
    "                break\n",
    "\n",
    "    # search up to 5 downstream neighbours\n",
    "    for i in range(lncRNA_index + 1, len(features)):\n",
    "        feature = features[i]\n",
    "        if feature['type'] == 'gene':\n",
    "            last_gene_downstream = i\n",
    "        elif feature['type'] == 'mRNA' and last_gene_downstream is not None:\n",
    "            downstream_mRNAs.append(i)\n",
    "            last_gene_downstream = None\n",
    "            if len(downstream_mRNAs) == num_neighbors:\n",
    "                break\n",
    "\n",
    "    return upstream_mRNAs[:num_neighbors], downstream_mRNAs[:num_neighbors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate distances and assess for overlapping\n",
    "def calculate_distance(lncRNA, mRNA):\n",
    "    if mRNA['end'] < lncRNA['start']:\n",
    "        return (lncRNA['start'] - mRNA['end']) / 1000\n",
    "    elif mRNA['start'] > lncRNA['end']:\n",
    "        return (mRNA['start'] - lncRNA['end']) / 1000\n",
    "    else:\n",
    "        overlap_bases = min(lncRNA['end'], mRNA['end']) - max(lncRNA['start'], mRNA['start']) + 1\n",
    "        return -overlap_bases / 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the output\n",
    "output_filename = os.path.splitext(input_filename)[0] + \"_lncker_output.txt\"\n",
    "\n",
    "with open(output_filename, 'w') as f_output:\n",
    "    # headers\n",
    "    f_output.write(\"Contig\\tlncRNA_ID\\tlncRNA_strand\\tmRNA_ID\\tmRNA_strand\\tDE_status\\tmRNA_Position\\tdistance_Kb\\tproduct\\n\")\n",
    "\n",
    "    # search for mRNA neighbours for the IDs of interest and verify their DE status\n",
    "    for gene_id in ids_interes:\n",
    "        for contig_id in contigs_interes[gene_id]:\n",
    "            features = contigs[contig_id]\n",
    "            for i, feature in enumerate(features):\n",
    "                if feature['type'] == 'lnc_RNA' and feature['metadata'].get('ID', '').endswith(gene_id):\n",
    "                    upstream_mRNAs, downstream_mRNAs = find_multiple_mRNA_neighbors(features, i)\n",
    "\n",
    "                    # process mRNAs upstream\n",
    "                    for idx, upstream_mRNA_idx in enumerate(upstream_mRNAs):\n",
    "                        upstream_mRNA = features[upstream_mRNA_idx]\n",
    "                        upstream_mRNA_id = upstream_mRNA['metadata'].get('ID', 'Unknown').lstrip(\"rna-\")\n",
    "                        upstream_mRNA_product = upstream_mRNA['metadata'].get('product', 'Unknown')\n",
    "                        distance_upstream = calculate_distance(feature, upstream_mRNA)\n",
    "                        de_status_upstream = \"DE\" if upstream_mRNA_id in de_mrnas else \"not DE\"\n",
    "\n",
    "                        # find overlapping genes\n",
    "                        distance_value = f\"{distance_upstream:.3f}\"\n",
    "\n",
    "                        f_output.write(f\"{contig_id}\\t{gene_id}\\t{feature['strand']}\\t{upstream_mRNA_id}\\t{upstream_mRNA['strand']}\\t{de_status_upstream}\\t{idx + 1}_U\\t{distance_value}\\t{upstream_mRNA_product}\\n\")\n",
    "\n",
    "                    # process mRNAs downstream\n",
    "                    for idx, downstream_mRNA_idx in enumerate(downstream_mRNAs):\n",
    "                        downstream_mRNA = features[downstream_mRNA_idx]\n",
    "                        downstream_mRNA_id = downstream_mRNA['metadata'].get('ID', 'Unknown').lstrip(\"rna-\")\n",
    "                        downstream_mRNA_product = downstream_mRNA['metadata'].get('product', 'Unknown')\n",
    "                        distance_downstream = calculate_distance(feature, downstream_mRNA)\n",
    "                        de_status_downstream = \"DE\" if downstream_mRNA_id in de_mrnas else \"not DE\"\n",
    "\n",
    "                        # find overlapping genes\n",
    "                        distance_value = f\"{distance_downstream:.3f}\"\n",
    "\n",
    "                        f_output.write(f\"{contig_id}\\t{gene_id}\\t{feature['strand']}\\t{downstream_mRNA_id}\\t{downstream_mRNA['strand']}\\t{de_status_downstream}\\t{idx + 1}_D\\t{distance_value}\\t{downstream_mRNA_product}\\n\")\n",
    "\n",
    "print(f\"Results have been written to {output_filename}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
